CrossrefReferenceLinkingPlugin.php

```diff
@@ -0,0 +1,474 @@
+		$preliminaryOutput =& $params[0];
+		$request = Application::get()->getRequest();
+		$context = $request->getContext();
+		// Crossref export cannot be executed via CLI any more, thus there will always be a context
+		$contextId = $context->getId();
+		$citationDao = DAORegistry::getDAO('CitationDAO'); /** @var CitationDAO $citationDao */
+
+		$rfNamespace = 'http://www.crossref.org/schema/5.3.1';
+		$articleNodes = $preliminaryOutput->getElementsByTagName('journal_article');
+		foreach ($articleNodes as $articleNode) {
+			$doiDataNode = $articleNode->getElementsByTagName('doi_data')->item(0);
+			$doiNode = $doiDataNode->getElementsByTagName('doi')->item(0);
+			$doiValue = $doiNode->nodeValue;
+			// There should be only one DOI
+			/** @var Doi $doi */
+			$doi = Repo::doi()->getCollector()->filterByContextIds([$contextId])->filterByIdentifier($doiValue)->getMany()->first();
+			if (!$doi) return false;
+			$publications = Repo::publication()->getCollector()->filterByDoiIds([$doi->getId()])->getMany();
+			if ($publications->count() < 1) return false;
+			$submission = Repo::submission()->get($publications->first()->getData('submissionId'));
+			$articleCitations = $citationDao->getByPublicationId($submission->getCurrentPublication()->getId());
+			$articleCitationsArray = $articleCitations->toArray();
+			if (!empty($articleCitationsArray)) {
+				$citationListNode = $preliminaryOutput->createElementNS($rfNamespace, 'citation_list');
+				foreach($articleCitationsArray as $citation) {
+					$rawCitation = $citation->getRawCitation();
+					if (!empty($rawCitation)) {
+						$citationNode = $preliminaryOutput->createElementNS($rfNamespace, 'citation');
+						$citationNode->setAttribute('key', $citation->getId());
+						// if Crossref DOI already exists for this citation, include it
+						// else include unstructred raw citation
+						if ($citation->getData($this->getCitationDoiSettingName())) {
+							$citationNode->appendChild($node = $preliminaryOutput->createElementNS($rfNamespace, 'doi', htmlspecialchars($citation->getData($this->getCitationDoiSettingName()), ENT_COMPAT, 'UTF-8')));
+						} else {
+							$citationNode->appendChild($node = $preliminaryOutput->createElementNS($rfNamespace, 'unstructured_citation', htmlspecialchars($rawCitation, ENT_COMPAT, 'UTF-8')));
+						}
+						$citationListNode->appendChild($citationNode);
+					}
+				}
+				$doiDataNode->parentNode->insertBefore($citationListNode, $doiDataNode->nextSibling);
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * During the article DOI registration with Crossref, get the citations diagnostic ID from the Crossref response.
+	 *
+	 * @param $hookName string Hook name 'crossrefexportplugin::deposited'
+	 * @param $params array [
+	 *  @option CrossrefExportPlugin
+	 *  @option string XML reposonse from Crossref deposit
+	 *  @option Submission
+	 * ]
+	 */
+	public function getCitationsDiagnosticId(string $hookName, array $params): bool
+	{
+		$response = & $params[1];
```

Comment:
I think it's good to add some type hinting to these unknown variables.
Then, instead of using `/** @var */` maybe this syntax https://phpstan.org/writing-php-code/phpdoc-types#array-shapes together with a good IDE extension, might work (or not) 🤔

Reply:
Hmmm... I added the `/**  @var */` -- the other option did not work with my current IDE...